classes/cloner.php

```diff
@@ -0,0 +1,159 @@
+     */
+    public static function queue($data, core_course_category $src, core_course_category $dest) {
+        $courseids = array_keys($src->get_courses(['recursive' => false]));
+
+        foreach ($courseids as $courseid) {
+            $task = new clone_course_task();
+            $task->set_custom_data((object) [
+                'courseid' => $courseid,
+                'destid' => $dest->id,
+                'srcid' => $src->id,
+                'data' => $data
+            ]);
+            manager::queue_adhoc_task($task, true);
+        }
+    }
+
+    /**
+     * Performs the cloning of the given course to the destination category.
+     * This is intended to be called from within an adhoc task.
+     * @param int $courseid course to clone (in $src category)
+     * @param core_course_category $src source category
+     * @param core_course_category $dest destination category
+     * @param object $data form data
+     */
+    public static function clone_course(int $courseid, core_course_category $src, core_course_category $dest, object $data) {
+        global $DB;
+
+        core_php_time_limit::raise(600);
+
+        $course = get_course($courseid);
+        $idn = explode('_', $course->shortname);
+        $shortname = reset($idn) . '_' . $dest->idnumber;
+
+        // If a course matching the shortname and destination category already exists, skip it.
+        if ($DB->record_exists("course", ["shortname" => $shortname, "category" => $dest->id])) {
+            throw new Exception("Course with shortname {$shortname} already exists in the category. Skipped.");
+        }
+
+        $options = [
+            ['name' => 'activities', 'value' => 1],
+            ['name' => 'blocks', 'value' => 1],
+            ['name' => 'filters', 'value' => 1],
+            ['name' => 'users', 'value' => 0],
+            ['name' => 'role_assignments', 'value' => 0],
+            ['name' => 'comments', 'value' => 0],
+            ['name' => 'userscompletion', 'value' => 0],
+            ['name' => 'logs', 'value' => 0],
+            ['name' => 'grade_histories', 'value' => 0]
+        ];
+
+        $clone = core_course_external::duplicate_course($course->id, $course->fullname, $shortname, $dest->id, 0, $options);
+
+        $newid = $clone['id'];
+        $newshortname = $clone['shortname'];
+        $newfullname = str_replace($src->idnumber, $dest->idnumber, $course->fullname);
+
+        $DB->set_field_select('course', 'fullname', $newfullname, "id = ?", [$newid]);
+        $DB->set_field_select('course', 'startdate', $data->startdate, "id = ?", [$newid]);
+        $DB->set_field_select('course', 'enddate', $data->enddate, "id = ?", [$newid]);
+
+        $entry = "Cloned {$course->id}/{$course->shortname} into {$newid}/{$newshortname};";
+
+        // Log success to event log.
+        self::log_clone_status($entry, $clone['id']);
+
+        // Also log to cron logs.
+        mtrace($entry);
```

Comment:
This will only mtrace if its a success but not failures. I've mode the mtrace to inside log_clone_status so be consistent and simpler